### DISCUSSION TITLE
What do static vectors do during training?

### DISCUSSION THREAD

**Discussion by BramVanroy** [explosion/spaCy#8709] [Help: Other Questions] [answered]
I understand that you can add [static vectors](https://spacy.io/usage/embeddings-transformers#static-vectors) to a trainable component, e.g.

```
vectors = "path/to/vectors"

[components.ner.model.tok2vec.embed]
@architectures = "spacy.MultiHashEmbed.v2"
width = 96
include_static_vectors = true
```

So my question is, what do these vectors _do_, or how are they used during training? Are the static vectors passed as additional input features for the NER model alongside the word? The same question goes for when we use `include_static_vectors = true` on a Tok2Vec layer. 

So I am trying to understand the interplay between static vectors and other components. An additional example: [here](https://spacy.io/usage/linguistic-features#vectors-similarity) we read that similarity scores can still be calculated by using smaller models without static vectors. That raise the question: which component _sets_ the vectors that are responsible for the similarity?

**Comment by polm** [comment, marked as answer]
> So my question is, what do these vectors do, or how are they used during training?

Static vectors are used for one of the [pretraining objectives][objectives] called "LMAO" where the model attempts to reproduce the vectors. Also as indicated in the [static vectors docs](https://spacy.io/usage/embeddings-transformers#static-vectors) they can be used as a feature in the tok2vec representation. 

[objectives]: https://spacy.io/usage/embeddings-transformers#pretraining-objectives

> That raise the question: which component sets the vectors that are responsible for the similarity?

As [indicated in the docs](https://spacy.io/api/doc#similarity) similarity uses an average of word vectors by default (the non-default cause is a user hook to override this). The vector value is just `Doc.vector` (or Span or Token). That value uses word vectors if available, and if not (as in the small models), falls back to the tok2vec representation, which is not tuned for similarity.

The implementations of each of these functions is quite small, if you're curious I would recommend looking at the source.

**Reply by BramVanroy** [reply]
Alright, so after some digging I found what I was looking for and it is indeed a simple concatenation exercise.

https://github.com/explosion/spaCy/blob/8a95475b3dce2e52bc9be53a7b8c9ad49d7fc32c/spacy/ml/models/tok2vec.py#L161-L173

@polm So if I understand correctly, if static vectors are included then each Token will have a vector representation based on that context-insensitive table. If such table is not present, then the context-sensitive output of the Tok2Vec component will be used to set Token.vector - assuming that such a component exists. Can you also, then, get the context-sensitive vector representation even if static vectors are available? Context-insensitive are useful for similarity, but context-sensitive representations are useful for other tasks, too.

**Reply by polm** [reply]
> Can you also, then, get the context-sensitive vector representation even if static vectors are available?

Yes, it's just the mean of the `tensor` attribute. 

https://github.com/explosion/spaCy/blob/2a8eeed5da8bcc169f6c24459f67f49059c77181/spacy/tokens/doc.pyx#L632,L634

**Reply by adrianeboyd** [reply]
The standalone `tok2vec` component sets `doc.tensor`, the built-in tok2vec layers do not, so you end up with the tensor from the last `tok2vec` in the pipeline.

**Reply by BramVanroy** [reply]
@polm I think you misunderstood. My question was: can you get both the context-sensitive representation (from the last `tok2vec`/`tranformers` layer as @adrianeboyd says) **and** the context-insensitive representation (from static vectors, if part of pipeline), but I have figured it out with your help.

https://github.com/explosion/spaCy/blob/2a8eeed5da8bcc169f6c24459f67f49059c77181/spacy/tokens/token.pyx#L399-L413

CMIIW, I think these are the if-statements:

1. Get the vector from a user_defined_hook if such a hook is implemented
2. Get slice from doc vector if no static vectors exist in Vocab (the same Token.tensor)
3. Get static vector from Vocab store if it exists

So as you said, and in summary for myself when I forgot in the future:

- `Token.vector` is the context-insensitive representation from static vector if available (e.g. in larger models) 
- if static vectors are not available (smaller models), then `Token.vector == Token.tensor` 
- `Token.similarity` uses `Token.vector`, so if static vectors are not available similarity will be calculated on `Token.tensor` (context-sensitive) which may give unexpected results
- if you really want to you can customize similarity calculations in a [user hook](https://spacy.io/usage/processing-pipelines#custom-components-user-hooks)
